### PR TITLE
refactoring of mlperceptron

### PR DESCRIPTION
**Pull request by temporaer**
I tried to aim for a few goals as discussed in the sklearn pull request:
- have different concepts=classes for activation functions and weights
- have different concepts for weight updates
- not to sacrifice speed for doing the above
- potentially allow to reuse components for multiple layers

I'm pretty sure I removed some bugs in the sequence of gradient calculation and weight updates in the process. At the same time, I 'simplified' few features away while debugging (e.g. the special tanh), but I'll put them in again once we can agree on the structure.

### CODE REVIEW

**Comment by larsmans** [comment]
Thanks heaps, I'll try to do a proper review this week (busy with research, but this might actually come in handy). Results might be bad because there's no regularization. I'd like to include at least L2 regularization.

**Comment by temporaer** [comment]
not sure what you mean with the L2 regularization, it is included. When you do the review, can you elaborate a bit on how you imagined reusing cost functions from SGD? I left them in the main loop for now.

**Comment by larsmans** [comment]
I thought I'd disabled regularization in the Python wrapper. Let's get back to this when I really have some time to sparse. Feel free to hack further in the meantime (in a separate branch).

Final remark: I don't think we can actually reuse much code from SGD, because it does OvA training only and I want true multiclass training here. But we can implement the same functions and architecture so we get e.g. hinge loss networks.

File: sklearn/neural/backprop_sgd.pyx [hunk 1]
```diff
@@ -289,38 +446,26 @@ def backprop_sgd(self, X, np.ndarray Y):
                 end = n_samples
                 start = n_samples - batchsize
 
+
+            # densify outputs
```
Comment (larsmans) [inline sklearn/neural/backprop_sgd.pyx]: What is the purpose of this? `sparse` is true iff `X` is sparse; `Y` is always dense.

File: sklearn/neural/backprop_sgd.pyx [hunk 2]
```diff
@@ -179,15 +219,147 @@ cdef double log_loss_multiclass(double[:, ::1] output, double[:, ::1] T):
...
+        self.add_bias(out)
+
+    cdef inline void derivative_X_dense(self, inp, out):
+        np.dot(inp, self.coef, out=out)
+
+    cdef inline void derivative_W_dense(self, double[:, :] delta, double[:, :] act):
+        np.dot(delta.T, act, out=self.coef_grad)
+        np.sum(self.coef_grad, axis=1, out=self.bias_grad)
+
+cdef class WeightLayerSparse(WeightLayer):
```
Comment (larsmans) [inline sklearn/neural/backprop_sgd.pyx]: I'm going to remove this class; since it has no data attributes, the methods can go into `WeightLayer`.

File: sklearn/neural/backprop_sgd.pyx [hunk 3]
```diff
@@ -197,87 +369,72 @@ def backprop_sgd(self, X, np.ndarray Y):
...
-    bias_output_grad = np.empty(n_targets, dtype=np.float64)
+    Y_batch = np.empty((batchsize, n_targets), dtype=np.float64)
 
-    # velocities for momentum method
-    v_hidden = np.zeros(self.coef_hidden_.shape)
-    v_output = np.zeros(self.coef_output_.shape)
-    v_bias_hidden = np.zeros(self.intercept_hidden_.shape)
-    v_bias_output = np.zeros(self.intercept_output_.shape)
+    if sparse:
+        rowind = np.arange(n_samples, dtype=np.int32)
```
Comment (larsmans) [inline sklearn/neural/backprop_sgd.pyx]: Why is minibatch learning disabled for dense input?

**Comment by larsmans** [comment]
@temporaer Can you still run the document classification example? It seems to be eating up all the CPU without doing anything (`verbose=True` but no output after several minutes waiting).

File: sklearn/neural/backprop_sgd.pyx (hunk 1 above)
Comment (temporaer) [inline sklearn/neural/backprop_sgd.pyx]: In the sparse implementation, randomization is over instances in the training set, in the dense version, randomization is over batches. The loops collect the different rows of Y into a dense matrix. 

I believe it should be enough to randomize batches, but I did not want to break your functionality.

File: sklearn/neural/backprop_sgd.pyx (hunk 2 above)
Comment (temporaer) [inline sklearn/neural/backprop_sgd.pyx]: I'd prefer it if both classes could offer the same interface instead, so the caller does not need to distinguish between them. What do you think about having different classes for `X` (sparse/not sparse) and letting the weight layer choose what to do with it?

File: sklearn/neural/backprop_sgd.pyx (hunk 3 above)
Comment (temporaer) [inline sklearn/neural/backprop_sgd.pyx]: it shouldn't be, unless I missed something. See my first comment above.

**Comment by temporaer** [comment]
The problem seems to be the sparse derivative for W0. I didn't see this, since i tested with dense inputs...looking into this.

File: sklearn/neural/backprop_sgd.pyx (hunk 1 above)
Comment (larsmans) [inline sklearn/neural/backprop_sgd.pyx]: I see. But that makes the code quite hard to read, so I suggest we pick one strategy. `SGDClassifier` randomizes samples, so I picked that strategy. I can't immediately see the consequences of shuffling the batches only; wouldn't that take longer to converge if the samples are not pre-shuffled?

File: sklearn/neural/backprop_sgd.pyx (hunk 2 above)
Comment (larsmans) [inline sklearn/neural/backprop_sgd.pyx]: You mean `cdef` classes `DenseArray` and `CSRMatrix`? I think that's overengineering as there are only two places where the distinction matters: when computing the input->hidden activations, and their gradients. Also, if we do want to get more hidden layers in the future, then each hidden layer will be doing a useless check for sparse input.

In the long run, we might refactor the `SequentialDataset` from `sklearn.utils` to work with this class, but I'm leaving that out of scope as it causes a lot of changes in `SGDClassifier` as well.

File: sklearn/neural/backprop_sgd.pyx (hunk 1 above)
Comment (temporaer) [inline sklearn/neural/backprop_sgd.pyx]: It is a bit messy, agreed. I chose randomization over batches because of the speed improvement (only views are required, no looping). It is probably a good idea to pre-shuffle your data before applying this technique, but I do not expect a strong effect on convergence otherwise.

**Comment by larsmans** [comment]
Ok. I like your way of structuring the code, btw. Sorry that it took me so long to do a review.

**Comment by larsmans** [comment]
The dense case seems broken too: I get a `dtype` mismatch because `Y_batch` must be `np.float64`, while `Y` is `np.int`. I'm writing unit tests.

File: sklearn/neural/backprop_sgd.pyx (hunk 1 above)
Comment (larsmans) [inline sklearn/neural/backprop_sgd.pyx]: Right, but we can't expect users to pre-shuffle their inputs -- not all users are experts (and honestly, I tend to forget that kind of stuff all the time myself).

Comment (temporaer) [inline sklearn/neural/backprop_sgd.pyx]: Since this is a convenience/speed tradeoff... what about adding an classifier `shuffle=True` param, so that the classifier by default shuffles the (dense) data before learning?

Comment (larsmans) [inline sklearn/neural/backprop_sgd.pyx]: How much is the speed difference? I'm not too fond of messing with `X` in-place, nor of copying it.
